presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2515,42 +2534,274 @@ private void analyzeColumnMask(String currentIdentity, Table table, QualifiedObj
+                                }
 
-                String name = withQuery.getName().getValue().toLowerCase(ENGLISH);
-                if (withScopeBuilder.containsNamedQuery(name)) {
-                    throw semanticException(DUPLICATE_NAMED_QUERY, withQuery, "WITH query name '%s' specified more than once", name);
-                }
+                                List<Type> anchorFieldTypes = anchorType.getVisibleFields().stream()
+                                        .map(Field::getType)
+                                        .collect(toImmutableList());
+                                List<Type> stepFieldTypes = stepType.getVisibleFields().stream()
+                                        .map(Field::getType)
+                                        .collect(toImmutableList());
+                                if (!anchorFieldTypes.equals(stepFieldTypes)) {
+                                    analysis.addRelationCoercion(step, anchorFieldTypes.toArray(Type[]::new));
+                                }
+
+                                for (int i = 0; i < anchorFieldTypes.size(); i++) {
+                                    if (!typeCoercion.canCoerce(stepFieldTypes.get(i), anchorFieldTypes.get(i))) {
+                                        throw semanticException(
+                                                TYPE_MISMATCH,
+                                                step,
+                                                "recursion step relation output type (%s) is not coercible to recursion base relation output type (%s) at column %s",
+                                                stepFieldTypes.get(i),
+                                                anchorFieldTypes.get(i),
+                                                i + 1);
+                                    }
+                                }
 
-                // check if all or none of the columns are explicitly alias
-                if (withQuery.getColumnNames().isPresent()) {
-                    List<Identifier> columnNames = withQuery.getColumnNames().get();
-                    RelationType queryDescriptor = analysis.getOutputDescriptor(query);
-                    if (columnNames.size() != queryDescriptor.getVisibleFieldCount()) {
-                        throw semanticException(MISMATCHED_COLUMN_ALIASES, withQuery, "WITH column alias list has %s entries but WITH query(%s) has %s columns", columnNames.size(), name, queryDescriptor.getVisibleFieldCount());
+                                analysis.setScope(withQuery.getQuery(), aliasedAnchorScope);
+                                analysis.registerExpandableQuery(withQuery.getQuery(), recursiveReference);
+                                withScopeBuilder.withNamedQuery(name, withQuery);
+                                continue;
```

Comment:
The many levels of nesting and this `continue` here make it hard to follow the flow of the code. One possible way of reorganizing this is to move this whole section to a new method and have it return "true" if it was able to handle the recursive query. Then, the code below would do:

```
if (!handled) {
   // 2. 
   ...
   // process plain query
   ...
}
```

Let me play with the code a bit to see if I can come up with something better.

Reply:
Also, instead of splitting the method into two sections based on `with.isRecursive()`, it seems we should be able to structure in a similar manner to this and avoid a bit of duplication:

```java
With with = node.getWith().get();
Scope.Builder withScopeBuilder = scopeBuilder(scope);

for (WithQuery withQuery : with.getQueries()) {
    String name = withQuery.getName().getValue().toLowerCase(ENGLISH);
    if (withScopeBuilder.containsNamedQuery(name)) {
        throw semanticException(DUPLICATE_NAMED_QUERY, withQuery, "WITH query name '%s' specified more than once", name);
    }

    boolean isRecursive = false;
    if (with.isRecursive()) {
        isRecursive = tryProcessRecursiveQuery(...);
    }

    if (!isRecursive) {
        // check if all or none of the columns are explicitly alias
        Query query = withQuery.getQuery();
        if (withQuery.getColumnNames().isPresent()) {
            validateColumnAliases(withQuery.getColumnNames().get(), analysis.getOutputDescriptor(query).getVisibleFieldCount());
        }

        process(query, withScopeBuilder.build());
        withScopeBuilder.withNamedQuery(name, withQuery);
    }
}

Scope withScope = withScopeBuilder.build();
analysis.setScope(with, withScope);
return withScope;
```

Possibly, even this for the inner condition, but I'm not a big fan of the side-effect in the `if` condition:

```java
...
if (!with.isRecursive() || !tryProcessRecursiveQuery(...)) {
   ...
}
...
```